libraries/chain/block_state.cpp

```diff
@@ -102,5 +102,34 @@ namespace eosio::chain {
    ,_cached_trxs( std::move(trx_metas) )
    {}
 #endif
+
+   std::optional<qc_data_t> block_state::get_best_qc() const {
+      auto block_number = block_num();
+
+      // if pending_qc does not have a valid QC, consider valid_qc only
+      if( !pending_qc.is_quorum_met() ) {
+         if( valid_qc ) {
+            return qc_data_t{ quorum_certificate{ block_number, *valid_qc },
+                              qc_info_t{ block_number, valid_qc->is_strong() }};
+         } else {
+            return std::nullopt;;
+         }
+      }
+
+      // extract valid QC from pending_qc
+      valid_quorum_certificate valid_qc_from_pending(pending_qc);
```

Comment:
I don't like that we use a constructor that has an `assert(0)` inside of it if it is called incorrectly. I get that it shouldn't happen in this case since you have already ensured that by this point `pending_qc.is_quorum_met() == true`. But this is a fragile pattern. It would be better to have a member function in `pending_quorum_certificate` to extract an optional `valid_quorum_certificate`.

Anyway, this is just a note for the future. I don't expect this to change in this PR.

Reply:
Added a TODO warning.